[FlexAttention] Support training bias for eager (#136910)

drisspg opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #137526

This PR is Part 2 of the implementation started in https://github.com/pytorch/pytorch/pull/136910, rolled in the updates from https://github.com/pytorch/pytorch/pull/137451. Original was reverted due to calls to #@torch.libary at `import torch` time, so added a call to register at first call to `ModIndex`



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @rec

drisspg committed:
[FlexAttention] Support training bias for eager (#136910)

Part 2 of https://github.com/pytorch/pytorch/pull/136910

With some updates from: https://github.com/pytorch/pytorch/pull/137451
And updates to lib name + lazy registration

[ghstack-poisoned]

vadimkantorov commented:
@drisspg I wonder if FlexAttention could be added as a backend for existing SDPA method, maybe then it would be simpler for existing SDPA workloads to switch to it - for benchmarking flex attention in the basic setup (without score mods etc)...

drisspg committed:
Update on "[FlexAttention] Support training bias for eager (#136910)"


Part 2 of https://github.com/pytorch/pytorch/pull/136910

With some updates from: https://github.com/pytorch/pytorch/pull/137451
And updates to lib name + lazy registration

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

drisspg committed:
Update on "[FlexAttention] Support training bias for eager (#136910)"


Part 2 of https://github.com/pytorch/pytorch/pull/136910

With some updates from: https://github.com/pytorch/pytorch/pull/137451
And updates to lib name + lazy registration

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

drisspg committed:
Update on "[FlexAttention] Support training bias for eager (#136910)"


Part 2 of https://github.com/pytorch/pytorch/pull/136910

With some updates from: https://github.com/pytorch/pytorch/pull/137451
And updates to lib name + lazy registration

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

drisspg committed:
Update on "[FlexAttention] Support training bias for eager (#136910)"


Part 2 of https://github.com/pytorch/pytorch/pull/136910

With some updates from: https://github.com/pytorch/pytorch/pull/137451
And updates to lib name + lazy registration

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

drisspg commented on torch/_higher_order_ops/flex_attention.py:
```diff
@@ -69,6 +69,79 @@ def _permute_strides(out: torch.Tensor, query_strides: Tuple[int, ...]) -> torch
+
+    @staticmethod
+    def forward(x: Tensor, indices: List[Tensor]) -> Tensor:
+        if not ModIndex._is_registered:
```
Do I need to add any locks around this to avoid any threading issues?

drisspg commented:
@vadimkantorov  there has been some talk abou tthis and it might make sense if sdpa is called under compile, feel free to create a separate issue for this

drisspg committed:
Update on "[FlexAttention] Support training bias for eager (#136910)"


This PR is Part 2 of the implementation started in https://github.com/pytorch/pytorch/pull/136910, rolled in the updates from https://github.com/pytorch/pytorch/pull/137451. Original was reverted due to calls to #torch.libary at `import torch` time, so added a call to register at first call to `ModIndex`



[ghstack-poisoned]

drisspg committed:
Update on "[FlexAttention] Support training bias for eager (#136910)"


This PR is Part 2 of the implementation started in https://github.com/pytorch/pytorch/pull/136910, rolled in the updates from https://github.com/pytorch/pytorch/pull/137451. Original was reverted due to calls to #torch.libary at `import torch` time, so added a call to register at first call to `ModIndex`



[ghstack-poisoned]

drisspg commented on test/inductor/test_flex_attention.py:
```diff
@@ -989,14 +989,15 @@ def composed_score_mod(score, b, h, m, n):
         self.run_test(composed_score_mod, dtype)
 
     @supported_platform
+    # @expectedFailure  # TODO: Remove this after supporting compiled flex attention with training bias
```
I dont really trust this...

Chillee replied:
You don't believe in expectedFailure?

drisspg replied:
This is why I didnt trust this:

https://github.com/pytorch/pytorch/pull/137526/files#diff-e3963412cc249e81fecfcf8774f5428b2b5e837ff3633ae13d8b7886ab5bc3b9R2074

drisspg committed:
Update on "[FlexAttention] Support training bias for eager (#136910)"


This PR is Part 2 of the implementation started in https://github.com/pytorch/pytorch/pull/136910, rolled in the updates from https://github.com/pytorch/pytorch/pull/137451. Original was reverted due to calls to #torch.libary at `import torch` time, so added a call to register at first call to `ModIndex`



[ghstack-poisoned]

drisspg committed:
Update on "[FlexAttention] Support training bias for eager (#136910)"


This PR is Part 2 of the implementation started in https://github.com/pytorch/pytorch/pull/136910, rolled in the updates from https://github.com/pytorch/pytorch/pull/137451. Original was reverted due to calls to #torch.libary at `import torch` time, so added a call to register at first call to `ModIndex`



[ghstack-poisoned]

drisspg committed:
Update on "[FlexAttention] Support training bias for eager (#136910)"


This PR is Part 2 of the implementation started in https://github.com/pytorch/pytorch/pull/136910, rolled in the updates from https://github.com/pytorch/pytorch/pull/137451. Original was reverted due to calls to #torch.libary at `import torch` time, so added a call to register at first call to `ModIndex`



[ghstack-poisoned]

drisspg committed:
Update on "[FlexAttention] Support training bias for eager (#136910)"


This PR is Part 2 of the implementation started in https://github.com/pytorch/pytorch/pull/136910, rolled in the updates from https://github.com/pytorch/pytorch/pull/137451. Original was reverted due to calls to #torch.libary at `import torch` time, so added a call to register at first call to `ModIndex`



[ghstack-poisoned]

vadimkantorov commented on torch/_higher_order_ops/flex_attention.py:
```diff
@@ -94,17 +91,23 @@ def _(
+        for idx, idx_indim in zip(indices, indices_indims):
+            # The index is not a being batched, we should unsqueeze and expand to val
+            if idx_indim is None:
+                expanded_indices.append(idx.unsqueeze(0).expand(value.shape))
```
could simply `.append(idx.expand_as(value))` be used instead?

also maybe explicit `unsqueeze(0)` could be omitted? as broadcasting should already do it

drisspg replied:
Good tip, thanks

drisspg committed:
Update on "[FlexAttention] Support training bias for eager (#136910)"


This PR is Part 2 of the implementation started in https://github.com/pytorch/pytorch/pull/136910, rolled in the updates from https://github.com/pytorch/pytorch/pull/137451. Original was reverted due to calls to #torch.libary at `import torch` time, so added a call to register at first call to `ModIndex`



[ghstack-poisoned]

drisspg committed:
Update on "[FlexAttention] Support training bias for eager (#136910)"


This PR is Part 2 of the implementation started in https://github.com/pytorch/pytorch/pull/136910, rolled in the updates from https://github.com/pytorch/pytorch/pull/137451. Original was reverted due to calls to #torch.libary at `import torch` time, so added a call to register at first call to `ModIndex`



[ghstack-poisoned]

drisspg committed:
Update on "[FlexAttention] Support training bias for eager (#136910)"


This PR is Part 2 of the implementation started in https://github.com/pytorch/pytorch/pull/136910, rolled in the updates from https://github.com/pytorch/pytorch/pull/137451. Original was reverted due to calls to #torch.libary at `import torch` time, so added a call to register at first call to `ModIndex`



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang rec

[ghstack-poisoned]

drisspg commented:
@pytorchbot merge

drisspg commented:
@mlazos verifeid that if I turn of _compile=False in the failing tst for the create_block_mask function error goes away, any ideas what might be happening here?

drisspg commented:
@mlazos w this modification to https://github.com/pytorch/pytorch/pull/137809 I get the tests to pass locally
```Shell
❯ g dft
torch/_dynamo/variables/misc.py --- Python
712 712         ctx = AutogradFunctionContextVariable.create(tx, args, kwargs)
713 713         args = [ctx, *args]
714 714         if isinstance(fn, types.FunctionType):
... 715             sig = inspect.signature(fn)
... 716             if len(args) - 1  == len(sig._parameters):
... 717                 args = args[1:]  # Don't use context
715 718             return variables.UserFunctionVariable(fn, source=source).call_function(
716 719                 tx, args, kwargs
717 720             )

```

drisspg committed:
Update

[ghstack-poisoned]

drisspg committed:
Update

[ghstack-poisoned]

drisspg commented:
@pytorchbot merge

drisspg commented:
@pytorchbot merge
